pkg/controller/remotemachineset/remotemachineset_controller.go

```diff
@@ -0,0 +1,502 @@
+
+	err := c.syncHandler(key.(string))
+	c.handleErr(err, key.(string))
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	logger := c.logger.WithField("clusterdeployment", key)
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		logger.Infof("error syncing cluster deployment: %v", err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("dropping cluster deployment out of the queue: %v", err)
+	c.queue.Forget(key)
+}
+
+func isMasterMachineSet(machineSet *cov1.ClusterMachineSet) bool {
+	return machineSet.NodeType == cov1.NodeTypeMaster
+}
+
+func (c *Controller) buildRemoteClusterClient(clusterDeployment *cov1.ClusterDeployment) (clusterapiclient.Interface, error) {
+	// Load the kubeconfig secret for communicating with the remote cluster:
+
+	// Step 1: Get the secret ref from the cluster
+	//
+	// TODO: The cluster controller should watch secrets until it
+	// sees a kubeconfig secret named after the cluster and then
+	// update cluster status with a reference to that
+	// secret. Until then we'll look for a
+	// "clustername-kubeconfig" secret.
+	//
+	// clusterConfig := cluster.Status.AdminKubeconfig
+	// if clusterConfig == nil {
+	// 	return nil, fmt.Errorf("unable to retrieve AdminKubeConfig from cluster status")
+	// }
+	secretName := clusterDeployment.Name + "-kubeconfig"
+
+	// Step 2: Retrieve secret
+	secret, err := c.kubeClient.CoreV1().Secrets(clusterDeployment.Namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	secretData := secret.Data["kubeconfig"]
+
+	// Step 3: Generate config from secret data
+	config, err := clientcmd.Load(secretData)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{})
+	restConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Step 4: Return client for remote cluster
+	return clusterapiclient.NewForConfig(restConfig)
+}
+
+// syncClusterDeployment will sync the clusterdeployment with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncClusterDeployment(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing clusterdeployment")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing clusterdeployment")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	clusterDeployment, err := c.clusterDeploymentLister.ClusterDeployments(namespace).Get(name)
+	if err != nil {
+		c.logger.Errorf("Error fetching clusterdeployment %v/%v", namespace, name)
+		return err
+	}
+
+	clusterapiCluster, err := c.clusterInformer.Clusters(clusterDeployment.Namespace).Get(clusterDeployment.Name)
```

Comment:
nit: no need to prefix cluster with 'clusterapi' since now we only have one type of Cluster